nilearn/connectome/rena_clustering.py

```diff
@@ -0,0 +1,623 @@
+
+    weights_unmasked = _compute_weights(masker, masked_data)
+
+    edges_unmasked = _make_3d_edges(vertices, is_mask=False)
+    edges_mask = _make_3d_edges(mask, is_mask=True)
+
+    # Apply mask to edges and weights
+    weights = np.copy(weights_unmasked[edges_mask])
+    edges = np.copy(edges_unmasked[:, edges_mask])
+
+    # Reorder the indices of the graph
+    max_index = edges.max()
+    order = np.searchsorted(np.unique(edges.ravel()), np.arange(max_index + 1))
+    edges = order[edges]
+
+    return edges, weights
+
+
+def weighted_connectivity_graph(masker, masked_data):
+    """ Creating symmetric weighted graph: data and topology are encoded by a
+    connectivity matrix.
+
+    Parameters
+    ----------
+    masker : NiftiMasker instance
+
+    masked_data : numpy array of shape [n_samples, n_features]
+        Image in brain space transformed into 2D data matrix
+
+    Returns
+    -------
+    connectivity : a sparse COO matrix
+        sparse matrix representation of the weighted adjacency graph
+    """
+    n_features = int(masker.mask_img_.get_data().sum())
+
+    edges, weight = _make_edges_and_weights(masker, masked_data)
+
+    connectivity = coo_matrix((weight, edges),
+                              (n_features, n_features)).tocsr()
+
+    # Making it symmetrical
+    connectivity = (connectivity + connectivity.T) / 2
+
+    return connectivity
+
+
+def _nn_connectivity(connectivity, threshold=1e-7):
+    """ Fast implementation of nearest neighbor connectivity
+
+    Parameters
+    ----------
+    connectivity : a sparse matrix in COOrdinate format.
+        sparse matrix representation of the weighted adjacency graph
+
+    threshold : float in the close interval [0, 1], optional (default 1e-7)
+        The treshold is setted to handle eccentricities.
+        In practice it is 1e-7.
+
+    Returns
+    -------
+    nn_connectivity : a sparse matrix in COOrdinate format.
+    """
+    n_features = connectivity.shape[0]
+
+    connectivity_ = coo_matrix(
+        (1. / connectivity.data, connectivity.nonzero()),
+        (n_features, n_features)).tocsr()
+
+    # maximum on the axis = 0
+    max_connectivity = connectivity_.max(axis=0).toarray()[0]
+    inv_max = dia_matrix((1. / max_connectivity, 0),
+                         shape=(n_features, n_features))
+
+    connectivity_ = inv_max * connectivity_
+
+    # Dealing with eccentricities, there are probably many neares neighbors
+    edge_mask = connectivity_.data > 1 - threshold
+
+    j_idx = connectivity_.nonzero()[1][edge_mask]
+    i_idx = connectivity_.nonzero()[0][edge_mask]
+
+    # Set weights to 1
+    weight = np.ones_like(j_idx)
+    edges = np.array([i_idx, j_idx])
+
+    nn_connectivity = coo_matrix((weight, edges), (n_features, n_features))
+
+    return nn_connectivity
+
+
+def _reduce_data_and_connectivity(labels, n_components, connectivity,
+                                  masked_data, threshold=1e-7):
```

Comment:
We should probably not have a default here

Reply:
It has been discussed in the previous PR I believe. Please see https://github.com/nilearn/nilearn/pull/1336#discussion_r92989879

IMO, this parameter is close to a precision parameter which as now real impact as soon as it is lower than a certain (low) value. Do you agree ? What do you suggest ?

Reply:
It has been discussed in the previous PR I believe. Please see https://github.com/nilearn/nilearn/pull/1336#discussion_r92989879

IMO, this parameter is close to a precision parameter which has no real impact as soon as it is lower than a certain (low) value. Do you agree ? What do you suggest ?

Reply:
@bthirion